packages/bundler-plugin-core/src/utils/getPreSignedURL.ts

```diff
@@ -25,14 +29,39 @@ export const getPreSignedURL = async ({
   uploadToken,
   serviceParams,
   retryCount = DEFAULT_RETRY_COUNT,
+  gitService,
 }: GetPreSignedURLArgs) => {
-  if (!uploadToken) {
-    red("No upload token found");
-    throw new NoUploadTokenError("No upload token found");
-  }
-
   const url = `${apiURL}/upload/bundle_analysis/v1`;
 
+  const headers = new Headers({
+    "Content-Type": "application/json",
+  });
+
+  const requestBody: RequestBody = serviceParams;
+  /**
+   * We currently require the branch to be in the format `owner:branch` to identify that it is a
+   * proper tokenless upload.
+   * See: https://github.com/codecov/codecov-api/pull/741
+   */
+  if (!uploadToken && serviceParams.branch?.includes(":")) {
+    if (gitService && gitService !== "") {
+      requestBody.git_service = gitService;
+    } else {
+      const foundGitService = findGitService();
+      if (!foundGitService || foundGitService === "") {
+        red("Failed to find git service for tokenless upload");
+        throw new UndefinedGitServiceError("No upload token provided");
+      }
+
+      requestBody.git_service = foundGitService;
+    }
+  } else if (uploadToken) {
+    headers.set("Authorization", `token ${uploadToken}`);
+  } else {
```

Comment:
these nested conditionals feel like a code smell to me, but I can't figure out a better way to address this so I wouldn't worry about it for now. Something to think about if we come back to this file